hazelcast/src/main/java/com/hazelcast/config/AbstractFactoryWithPropertiesConfig.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.config;
+
+import java.util.Objects;
+import java.util.Properties;
+
+/**
+ * Configuration base for config types with a factory class and its properties.
+ *
+ * @param <T> final child type
+ */
+public abstract class AbstractFactoryWithPropertiesConfig<T extends AbstractFactoryWithPropertiesConfig<T>> {
+
+    private boolean enabled;
+    private String factoryClassName;
+    private Properties properties = new Properties();
+
+    /**
+     * Returns the factory class name.
+     */
+    public String getFactoryClassName() {
+        return factoryClassName;
+    }
+
+    /**
+     * Sets the factory class name.
+     */
+    public T setFactoryClassName(String factoryClassName) {
```

Comment:
Can you also add some `@Nullable` or `@Nonnull` annotations to the parameters for reference? I guess it's not easy to determine in some cases as it depends on the concrete config class implementation.